File: Path_Sum_II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private ArrayList<ArrayList<Integer>> ret;
    public ArrayList<ArrayList<Integer>> pathSum(TreeNode root, int sum) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        ret = new ArrayList<ArrayList<Integer>>();
        if (root == null) return ret;
        solve (new ArrayList<Integer>(), root, sum);
        return ret;
    }
    
    public void solve(ArrayList<Integer> path, TreeNode root, int sum){
        if (root == null) return;
        ArrayList<Integer> newPath = new ArrayList<Integer>(path);
        newPath.add(root.val);
        sum -= root.val;
        if (root.left == null && root.right == null) {
            if (sum == 0) {
                ret.add(newPath);
            }
            else {
                return;
            }
        }
        solve(newPath, root.left, sum);
        solve(newPath, root.right, sum);
        
    }
}

File: C++/Path_Sum_II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector <int>> ret;
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (root == NULL){
            return ret;
        }
        vector<int> cand;
        ret.clear();
        cand.clear();
        solve(root, cand, sum);
        return ret;
    }
    void solve(TreeNode *root, vector<int> &cand, int sum){
        int sum_new = sum - root -> val;
        vector<int> cand_new(cand);
        cand_new.push_back(root -> val);
        if (root -> left == NULL && root -> right == NULL){
            if (sum_new == 0){
                ret.push_back(cand_new);
                return;
            }
            return;
        }
        if (root -> left != NULL) {
            solve(root -> left, cand_new, sum_new);
        }
        if (root -> right != NULL) {
            solve(root -> right, cand_new, sum_new);
        }
    }
};
